zerver/views/push_notifications.py

```diff
@@ -164,12 +158,55 @@ def self_hosting_auth_redirect(
         # Upload realm info and re-try. It should work now.
         send_server_data_to_push_bouncer(consider_usage_statistics=False)
         result = send_to_push_bouncer("POST", "server/billing", post_data)
-    except RemoteRealmServerMismatchError:  # nocoverage
-        return render(request, "zilencer/remote_realm_server_mismatch_error.html", status=403)
 
     if result["result"] != "success":  # nocoverage
         raise JsonableError(_("Error returned by the bouncer: {result}").format(result=result))
 
     redirect_url = result["billing_access_url"]
     assert isinstance(redirect_url, str)
+    return redirect_url
+
+
+@zulip_login_required
+@typed_endpoint
+def self_hosting_auth_redirect_endpoint(
+    request: HttpRequest,
+    *,
+    next_page: Optional[str] = None,
+) -> HttpResponse:
+    """
+    This endpoint is used by the web app running in the browser. We serve HTML
+    error pages, and in case of success a simple redirect to the remote billing
+    access link received from the bouncer.
+    """
+
+    user = request.user
+    assert user.is_authenticated
+    assert isinstance(user, UserProfile)
+
+    try:
+        redirect_url = self_hosting_auth_view_common(request, user, next_page)
+    except ResourceNotFoundError:
+        return render(request, "404.html", status=404)
+    except RemoteRealmServerMismatchError:
+        return render(request, "zilencer/remote_realm_server_mismatch_error.html", status=403)
```

Comment:
`# nocoverage` tags disappear in this refactor, because the new test covers these cases.

Reply:
Hmm, how can `zilencer/remote_realm_server_mismatch_error` be used here -- that code won't be present on a self-hosted Zulip server.

Reply:
@timabbott Oh yeah, good point. We added this in https://github.com/zulip/zulip/commit/ac8d5a5f0b9b8618d1609e672696361b02061eab#diff-c5ebe43c8cf276fb898525d41ffa701cf0a6bcc12022b25ec16bb24072d7e862 and I think the answer may simply be that this template should be moved into `zerver`?

Reply:
Yeah, that sounds right to me.

Reply:
Pushed a commit